res/controllers/lib/mididispatcher.mjs

```diff
@@ -0,0 +1,77 @@
+/**
+* MidiDispatcher routes incoming MIDI messages to registered callbacks. Register callbacks with
+* the setInputCallback method in your controller script module's init function and call
+* MidiDispatcher's receiveMidiData method in your controller script module's receiveData function.
+*
+* This class is not designed to handle System Exclusive or MIDI Clock messages. For those, implement logic
+* specific to your controller in the controller module's receiveData function before calling
+* MidiDispatcher.receiveData.
+*/
+export class MidiDispatcher {
+    /**
+     * @param {bool} noteOff - When setting the callback for a Note On message, also map the corresponding Note Off
+     * message to the same callback. Defaults to true.
+     */
+    constructor(noteOff) {
+        if (noteOff === undefined) {
+            noteOff = true;
+        }
+        this.noteOff = noteOff;
+        this.inputMap = new Map();
+    }
+    /**
+     * Set the callback for an incoming MIDI message.
+     * @param {Array} midiBytes - Array of numbers indicating the beginning of the MIDI messages. In most cases,
+     * this should be the first two bytes of the MIDI messages, for example [0x93, 0x27]
+     *
+     * Program change (starting with 0xC) and aftertouch (starting with 0xD) messages are distinguished by only
+     * their first byte, so in those cases the Array should only have one number.
+     *
+     * @param {function} callback - The callback that will be called by the receiveData method when a MIDI message
+     * matching midiBytes is received from the controller.
+     */
+    setInputCallback(midiBytes, callback) {
+        if (!Array.isArray(midiBytes)) {
+            throw new Error('MidiDispatcher.setInputCallback midiBytes must be an Array, received ' + midiBytes);
+        }
+        if (typeof midiBytes[0] !== 'number') {
+            throw new Error('MidiDispatcher.setInputCallback midiBytes must be an Array of numbers, received ' + midiBytes);
+        }
+        if (typeof callback !== 'function') {
+            throw new Error('MidiDispatcher.setInputCallback callback must be a function, received ' + callback);
+        }
+        // JavaScript is broken and believes [1,2] === [1,2] is false, so
+        // JSONify the Array to make it usable as a Map key.
+        const key = JSON.stringify(midiBytes);
```

Comment:
This PR enables the console extension so you can use `console.time` and `console.timeEnd` to profile the overhead. While I see that this solution certainly isn't the quickest, I'm not convinced to advocate changing it without actually seeing some numbers. Can you develop a test case that proves that this performance overhead actually becomes a bottleneck?

Reply:
Since we're only dealing with bytes, we could shift bytes to create a number. That would avoid string conversion.

Reply:
I could do this, but the result will be that a c++ solution can discard messages in no time while this solution will take some time.
Than the we keep discussing if that matters or not.

On an low-end hardware like a raspberry pi every cpu circle matters to allow glitch free usage of Mixxx.

I see zero issues to move the map into the c++ domain. Do you see an issue here?
The mapping will look exactly the same.

Reply:
What do you the think the js roundtrip will take?

Reply:
Starting a js callback is a big string conversion. What stands against to have the map in the c++ domain?

Reply:
So if I understand you correctly, your point is that  you want to port the MidiDispatcher to C++ for performance reasons, right? So when the controller receives a midi message, Mixxx would call the general entry point (`handleInput` iirc) just for that JS to call the C++ MidiDispatcher, which in turn then calls the JS again? I don't know enough about Qt and its V4 JIT capabilities to accurately assess that, but it doesn't seem very performant to me as well.

Reply:
Of cause not. 

I like to keep the MidiDispatcher modul, but cut out the handleMidiInput and inputMap.
setInputCallback should call someting like 
```engine.setInputMap(key, callback)```  instead of just ```this.inputMap.set(key, callback)```
handleMidiInput  in the C++ domain can than quickly discharge messages without the help of any V4 JIT.

Reply:
We have this pattern already for LEDs:
```
DJTechCDJ101.init = function(id) {
    engine.connectControl("[Channel1]", "play", "DJTechCDJ101.play_feedback_deck1");
    engine.connectControl("[Channel2]", "play", "DJTechCDJ101.play_feedback_deck2");
    engine.connectControl("[Channel1]", "cue_default", "DJTechCDJ101.cue_feedback_deck1");
    engine.connectControl("[Channel2]", "cue_default", "DJTechCDJ101.cue_feedback_deck2");
}
```
So lets adopt it here in the same way for midi messages.

Reply:
> Since we're only dealing with bytes, we could shift bytes to create a number. That would avoid string conversion.

This seems like a good idea. No changes to the public API of MidiDispatcher would be required. But I will not work on this now in this PR. This PR is past done. I have already moved on to #2920, which is also done. Optimizations can be done later. I will not overcomplicate this by porting a JS class to C++ without *evidence* of a practically relevant performance difference.

Reply:
Ah, it seems you missed some of the earlier discussions. We decided that we want a unified API where a controller gets passed a buffer with the Containing Bytes sent over the wire. No matter the protocol, Midi, hid, bulk, raw network traffic (for whatever reason, but might be useful). the Javascript InputMapper is just an utility class thats supposed to help eliminate some of the boilerplate code that comes with that approach for Midi data.

Reply:
I don't understand why this is mutable exclusive. You can still receive all telegrams, but we can release the JS engine form the normal boring workload. 

> This PR is past done 

Hey com on, that is a review. So demanding changes is normal. 

> I will not over complicate this by porting a JS class to C++ without evidence of a practically relevant performance difference.

The change is quite simple and the CPU saving is high, because a whole part of the stack is skipped. 
Isn't this a good evidence. Depending where do you measure you save 100 % CPU :-).

Reply:
> Hey com on, that is a review. So demanding changes is normal.

Sorry, I was very frustrated before. I have replaced the `JSON.stringify` with a hash function which only does some simple bitwise math. I think that should be faster than `JSON.stringify`. But I will not reimplement this in C++ without evidence there is a need for that.

Reply:
> I don't understand why this is mutable exclusive.

It isn't, but the approach would complicate the API design and create clutter (unless we just implement the MidiDispatcher but with the same API as it currently has, but that would create the several domain context switches again which wouldn't make sense performance-wise as well).

Reply:
> But I will not reimplement this in C++ without evidence there is a need for that.

What kind of evidence is acceptable? Even without a test it is obvious that it performs much better because we remove one layer of redirection. 

> It isn't, but the approach would complicate the API design and create clutter (unless we just implement the MidiDispatcher but with the same API as it currently has, but that would create the several domain context switches again which wouldn't make sense performance-wise as well).

It sounds like you probably did not get my point. The API from the mapping perspective will remain exactly the same, the number of context switches will remain in case of a processed message and will be reduced in all other cases.  
We just remove a layer of redirection. 

This PR: 
```
Controller::receive()
m_pEngine->handleInput()
executeFunction(m_handleInputFunction, args); // 1. Redirect + Context switch 
this.inputMap.get(key);
MidiDispatcher.handleMidiInput() // 2. Redirect 
    callback()
    or
    discharge // no context switch 
```

Proposal
```
Controller::receive()
m_inputMap.get(key);
    executeFunction(callback); // 1. Redirect + Context switch 
    or
    discharge // context switch required
```

Reply:
Controllers generally don't leave any MIDI inputs unmapped, so I don't think this would be worth the trouble.

Reply:
They have, some are sending sysex messages for some reasons.  

Isn't this also the foundation for our real time queue idea?

Reply:
If a controller is continuously sending sysex messages the script should probably be doing something with them.

> Isn't this also the foundation for our real time queue idea?

No. The idea with that is to insert a C++ object into the JS environment to provide an API to send jogwheel data from the script to the queue. This requires the script to implement whatever controller specific logic is needed.

Reply:
I also don't think that is trouble. We use this pattern for years: 
```
engine.connectControl("[Sampler" + extraParameter + "]", "play_indicator", "Hercules4Mx.onSamplerStateChange");
```

Reply:
I don't understand how that code snippet is relevant to this discussion?

Reply:
I agree with Be. If the controller sends the message, the software is usually supposed to act on it. The slight performance gain is not worth creating further entanglement on the C++ side.

Reply:
> I don't understand how that code snippet is relevant to this discussion?

Because that does the same, registers a js callback via an engine call. You was refereeing of trouble. 

The trouble is to move the map into the engine and add a API function to fill it. Done, no trouble but some CPU saved.

Reply:
We also don't limit the power of js in any way. It is still in charge to resister the callbacks and may also still listen to all messages.